gm_dotnet_managed/GmodNET/GameConsoleWriter.cs

```diff
@@ -13,9 +14,103 @@ internal class GameConsoleWriter : TextWriter
         public override string NewLine => "\n";
         public override Encoding Encoding => Encoding.UTF8;
 
-#pragma warning disable CA2101
-        [DllImport("tier0", CallingConvention = CallingConvention.Cdecl)]
-        private static extern void Msg([MarshalAs(UnmanagedType.LPUTF8Str)] string msg);
+        public GameConsoleWriter(ILua lua)
+        {
+            if (OperatingSystem.IsWindows())
+            {
+                Msg = Msg_Tier0;
+            }
+            else if (OperatingSystem.IsLinux())
+            {
+                lua.PushSpecial(SPECIAL_TABLES.SPECIAL_GLOB);
+                lua.GetField(-1, "game");
+                lua.GetField(-1, "IsDedicated");
+                lua.MCall(0, 1);
+                bool is_dedicated = lua.GetBool(-1);
+                lua.Pop(3);
+
+                if (is_dedicated)
+                {
+                    Msg = Msg_Tier0;
+                }
+                else
+                {
+                    Msg = Msg_Tier0_Client;
+                }
+            }
+            else if (OperatingSystem.IsMacOS())
+            {
+                unsafe
+                {
+                    IntPtr lib_handle = NativeLibrary.Load(Directory.GetCurrentDirectory() + "/GarrysMod_Signed.app/Contents/MacOS/libtier0.dylib");
+                    delegate* unmanaged[Cdecl]<byte*, void> msg_func = (delegate* unmanaged[Cdecl]<byte*, void>)NativeLibrary.GetExport(lib_handle, "Msg");
+
+                    Msg = (message) =>
+                    {
+                        byte* __message_gen_native = default;
+                        //
+                        // Setup
+                        //
+                        bool message__allocated = false;
+                        try
+                        {
+                            //
+                            // Marshal
+                            //
+                            if (message != null)
+                            {
+                                int message__bytelen = (message.Length + 1) * 3 + 1;
+                                if (message__bytelen > 260)
```

Comment:
Why we want to stop at exactly 260 for stack allocations?

Reply:
that's `DllImportGenerator`'s threshold.

> In the built-in system, marshalling a string contains an optimization for parameters passed by value to allocate on the stack (instead of through AllocCoTaskMem) if the string is below a certain length (MAX_PATH)